pkg/controller/networkpolicy/endpoint_querier.go

```diff
@@ -155,50 +188,150 @@ func (eq *endpointQuerier) QueryNetworkPolicies(namespace string, podName string
+}
+
+// processQueryResults processes the AppliedPolicies in endpoints, returns a set
+// of the policy UIDs, and manually inserts Kubernetes NetworkPolicy default isolation
+// rules if exists. The default isolation rule's direction depends on ifSource,
+// and has the index of -1 to indicate lower precedence.
+func processQueryResults(endpoints *EndpointNetworkPolicyRules, ifSource bool) (sets.Set[types.UID], []*RuleInfo) {
+	policyUIDs := sets.New[types.UID]()
+	isolation := make([]*RuleInfo, 0)
+	for _, internalPolicy := range endpoints.AppliedPolicies {
+		policyUIDs.Insert(internalPolicy.SourceRef.UID)
+		if internalPolicy.SourceRef.Type == controlplane.K8sNetworkPolicy {
+			// check if the Kubernetes NetworkPolicy creates ingress or egress isolation
+			for _, rule := range internalPolicy.Rules {
+				if rule.Direction == controlplane.DirectionIn && !ifSource {
+					isolation = append(isolation, &RuleInfo{Policy: internalPolicy, Index: -1,
+						Rule: &controlplane.NetworkPolicyRule{Direction: rule.Direction, Name: rule.Name, Action: rule.Action}})
+				} else if rule.Direction == controlplane.DirectionOut && ifSource {
+					isolation = append(isolation, &RuleInfo{Policy: internalPolicy, Index: -1,
+						Rule: &controlplane.NetworkPolicyRule{Direction: rule.Direction, Name: rule.Name, Action: rule.Action}})
+				}
+			}
+		}
+	}
+	return policyUIDs, isolation
+}
+
+// predictEndpointsRules returns the predicted rules effective from srcEndpoints to dstEndpoints.
+// Rules returned satisfy a. in source applied policies and destination egress rules,
+// or b. in source ingress rules and destination applied policies or c. applied to KNP default isolation.
+func predictEndpointsRules(srcEndpoints, dstEndpoints *EndpointNetworkPolicyRules) *RuleInfo {
+	commonRules := make([]*RuleInfo, 0)
+	if srcEndpoints != nil && dstEndpoints != nil {
+		srcPolicies, srcIsolated := processQueryResults(srcEndpoints, true)
+		dstPolicies, dstIsolated := processQueryResults(dstEndpoints, false)
+		for _, rule := range dstEndpoints.EgressRules {
+			if srcPolicies.Has(rule.Policy.SourceRef.UID) {
+				commonRules = append(commonRules, rule)
+			}
+		}
+		for _, rule := range srcEndpoints.IngressRules {
+			if dstPolicies.Has(rule.Policy.SourceRef.UID) {
+				commonRules = append(commonRules, rule)
+			}
+		}
+		for _, defaultDropRule := range srcIsolated {
+			commonRules = append(commonRules, defaultDropRule)
+		}
+		for _, defaultDropRule := range dstIsolated {
+			commonRules = append(commonRules, defaultDropRule)
+		}
+	}
+	// if no common rule or default isolation is found, return nil
+	if len(commonRules) == 0 {
+		return nil
+	}
+	// sort the common rules based on multiple closures, the top rule has the highest precedence
+	passRule := func(r1, r2 *RuleInfo) int {
+		if r2.Policy.SourceRef.Type != controlplane.K8sNetworkPolicy && r2.Policy.Rules[r2.Index].Action != nil && *r2.Policy.Rules[r2.Index].Action == v1beta1.RuleActionPass {
```

Comment:
This comparison is not symmetric, can it work? i.e. what if r1 is pass action? Also in terms of precedence, i think the logic here is wrong still... tierPriority, policyPriority and rulePriority should still take precedence

Reply:
Discussed offline and updated.